### ISSUE TITLE
Flutter_web navigation should provide a way to customize url strategy

### ISSUE THREAD

**Issue by rohitkumarluthra** [flutter/flutter#33245]
Hi Flutter Team,
I am from react background as I just started using the flutter web. I am trying to use the named routes in flutter web. For example I have 3 routes as shown in below code:
```
Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      initialRoute: '/',
      routes: <String, WidgetBuilder> { //5 
        '/': (context) => MyHomePage(title: 'Home Page'),
        '/first': (BuildContext context) => new FirstPage(), //6
        '/second' : (BuildContext context) => new SecondPage() //7
      },
    );
  }
```

**So there is a requirement:**
There should be a way to remove hash symbol(#) because currently it shows the url like `http://localhost:8080/#/`

**Comment by mkiisoft** [comment]
I have a walk-around using pushState:

`import 'dart:html' as html;`
`html.window.history.pushState(null, "Welcome", "/");`

inside your initState method, but we should have an option to remove it using some kind of WebScaffold or something like that.

**Comment by cic1988** [comment]
please provide a way to remove it. due to og:image I wanted to replace the meta tag content dynamically in index.html. with the hash it is not possible because the fragment won’t be sent to the server :(

**Comment by gagsy** [comment]
How I can use this in my main.dart page?
import 'dart:html' as html;

html.window.history.pushState(null, "Home", "/");

Please anyone can share the code or link

**Comment by ZeroCool00** [comment]
any solution yet? 

`ihtml.window.history.pushState(null, "Home", "/");`
if we try this then it is also removing parameter. so that is not good solution

**Comment by rlee1990** [comment]
Any updates on this? 
@mkiisoft how would we add this and how could we also add parameters?

**Comment by cicidev** [comment]
Any update on this ?

**Comment by Gorniv** [comment]
we need to work with a route like in the web... any updates?

**Comment by ci30** [comment]
Any update on this issue ?

**Comment by Ali-Elganzory** [comment]
Any updates?

**Comment by NaikSoftware** [comment]
Hello. Looks like regression after merge separate web repository into main flutter repository. Because I found that `window` object from `dart:ui` package is instance of `EngineWindow`, that has https://github.com/flutter/engine/blob/c0e29b6000a45fab05ba67a357c4787057f0fa55/lib/web_ui/lib/src/engine/window.dart#L120 setter for location strategy. But now cast to platform specific window is  impossible, because this class located in engine repository and not exposed in sdk. Also I found this comment, that describes how to set custom strategy https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/engine/browser_location.dart#L67 but now there no way to do it?

**Comment by zcfata** [comment]
I'm using this simple find and replace on my main.dart.js file

`sed -i.bak s/'"#"+a'/'""'/g ./build/web/main.dart.js`

and then using  window.history.pushState to change route

**Comment by 10686142** [comment]
Should we be implementing our own work-around or are there any plans for a possible solution in the foreseeable future?

**Comment by harshIos** [comment]
Hi all,
It's almost a year now when I reported this issue but as I can there is still no any solution available. So can we create some open source custom navigation library for flutter web same as available for react(https://reactnavigation.org/) by collaboration.

**Comment by RohitLuthra19** [comment]
Hi all,
It's almost a year now when I reported this issue but as I can there is still no any solution available. So can we create some open source custom navigation library for flutter web same as available for react(https://reactnavigation.org/) by collaboration.

**Comment by vivchar** [comment]
> `import 'dart:html' as html;`
> `html.window.history.pushState(null, "Home", "/");`
It works fine when I use in app navigation, but doesn't work when I open url without "#" in browser, example:
open `http://my-site.com/#/page1` - works
open `http://my-site.com/page1` - doesn't work

**Comment by rmill777** [comment]
Would love to see this fixed as well..

**Comment by tomrozb** [comment]
The main problem here is that https://domain.com/#/page links are not compatible with Android App Links.

So Android app doesn't understand https://domain.com/#/page link, and will not offer to open it in native app on Android device
Web Flutter app doesn't understand https://domain.com/page link, so it won't open correct page - it will display empty page (404).

Currently there's no way to create a deep link that works both on Flutter web and Android. That should be top priority for Flutter Web.

**Comment by edmbn** [comment]
But I think the main issue with it is that is broken. If you press back button on browser the url deletes /#/ and after that you can navigate normally but unable to use back button again in any route. The current history stack is well saved internally by flutter because appbar is managing it well, but when pressing back on browser it brokes.

**Comment by JeanChristophePal** [comment]
> https://domain.com/#/page

I managed to use deeplink with this one: 
https://domain.com/**%23**/page

**Comment by edmbn** [comment]
But why the browser back button is enabled when a new route es pushed and replaced? All the errors come from a miss click on back button when should be disabled. Then all new routes are never pushed to window.history so you can't go back from browser button. Everything working fine inside appbar back button.

**Comment by rmill777** [comment]
Currently, as a simple work around to @edmbn's issue, I wrap the HomePage container with 'WillPopScope(onWillPop: () async => false, child: $HomePageContainter$)' in the build method. This prevents users reaching the url without '/#/'..

**Comment by tomrozb** [comment]
> > Currently there's no way to create a deep link that works both on Flutter web and Android. That should be top priority for Flutter Web.
> 
> I managed to use deep link with **%23** instead of **#**. try this one:
> https://domain.com/%23/page

That doesn't work for me. Could you provide a valid link to show working example? 

In my case the link opens empty page on web.

**Comment by JeanChristophePal** [comment]
Unfortunately my Android app is not in production, only the Flutter PWA. If that can help one of my link is: https://yenboardgames.page.link/?link=https://yen-boardgames-helper.web.app/game/?gameId=kgh67F1VL2Qka4930xCi&apn=fr.yen.boardgames&afl=https://yen-boardgames-helper.web.app/%23/game/kgh67F1VL2Qka4930xCi&ofl=https://yen-boardgames-helper.web.app/%23/game/kgh67F1VL2Qka4930xCi

**Comment by JeanChristophePal** [comment]
Only the Flutter website is in prod. One share link is:
https://yenboardgames.page.link/?link=https://yen-boardgames-helper.web.app/game/?gameId=kgh67F1VL2Qka4930xCi&apn=fr.yen.boardgames&afl=https://yen-boardgames-helper.web.app/%23/game/kgh67F1VL2Qka4930xCi&ofl=https://yen-boardgames-helper.web.app/%23/game/kgh67F1VL2Qka4930xCi

**Comment by tomrozb** [comment]
@JeanChristophePal you mean that it works only when deployed to production? So for debug mode it displays empty page, and works fine in production?

**Comment by JeanChristophePal** [comment]
> @JeanChristophePal you mean that it works only when deployed to production? So for debug mode it displays empty page, and works fine in production?

No I mean you cannot test the dynamic link with the app installed the left branch of[ this schema](https://yenboardgames.page.link/?link=https://yen-boardgames-helper.web.app/home/?gameId=kgh67F1VL2Qka4930xCi&d=1&apn=fr.yen.boardgames&afl=https://yen-boardgames-helper.web.app/%23/home/kgh67F1VL2Qka4930xCi&ofl=https://yen-boardgames-helper.web.app/%23/home/kgh67F1VL2Qka4930xCi)

I fix the error of the above link. This new link should work better:

**Comment by tomrozb** [comment]
> > @JeanChristophePal you mean that it works only when deployed to production? So for debug mode it displays empty page, and works fine in production?
> 
> No I mean you cannot test this dynamic link with the android app installed the left branch of[ this schema](https://yenboardgames.page.link/?link=https://yen-boardgames-helper.web.app/home/?gameId=kgh67F1VL2Qka4930xCi&d=1&apn=fr.yen.boardgames&afl=https://yen-boardgames-helper.web.app/%23/home/kgh67F1VL2Qka4930xCi&ofl=https://yen-boardgames-helper.web.app/%23/home/kgh67F1VL2Qka4930xCi)
> 
> I fix an error of the previous link. This new link should work better:
> https://yenboardgames.page.link/?link=https://yen-boardgames-helper.web.app/home/?gameId=kgh67F1VL2Qka4930xCi&d=1&apn=fr.yen.boardgames&afl=https://yen-boardgames-helper.web.app/%23/home/kgh67F1VL2Qka4930xCi&ofl=https://yen-boardgames-helper.web.app/%23/home/kgh67F1VL2Qka4930xCi

Ok, so we're talking about two different things. I mentioned App Links (https://developer.android.com/training/app-links), whereas you're using Firebase Dynamic Links. Thanks for explanation, hope that helps somebody :)

**Comment by ArtikUA** [comment]
Here is a pull request with that feature: https://github.com/flutter/flutter/pull/55232

**Comment by jeeali** [comment]
Any Update!!!

**Comment by dhruvilp** [comment]
## A quick fix for it:
1. Go to `main.dart.js` and look for following line:
`H.a(window.location.pathname)+H.a(window.location.search):"#"+a},`
2. Remove **#** in above line
3. Result: 
`H.a(window.location.pathname)+H.a(window.location.search):""+a},`

This should give you https://example.com/ instead of <strike>https://example.com/#/</strike>

**Comment by ltvu93** [comment]
> ## A quick fix for it:
> 1. Go to `main.dart.js` and look for following line:
>    `H.a(window.location.pathname)+H.a(window.location.search):"#"+a},`
> 2. Remove **#** in above line
> 3. Result:
>    `H.a(window.location.pathname)+H.a(window.location.search):""+a},`
> 
> This should give you https://example.com/ instead of https://example.com/#/

I only change the url show in web page, but when we manually enter url to web, it not navigate.

**Comment by opringle** [comment]
Detailed guide on how to solve this issue - https://www.didierboelens.com/2020/02/flutter-web-404-direct-url-no-hash/

**Comment by danielmahon** [comment]
> The main problem here is that https://domain.com/#/page links are not compatible with Android App Links.
> 
> So Android app doesn't understand https://domain.com/#/page link, and will not offer to open it in native app on Android device
> Web Flutter app doesn't understand https://domain.com/page link, so it won't open correct page - it will display empty page (404).
> 
> Currently there's no way to create a deep link that works both on Flutter web and Android. That should be top priority for Flutter Web.

I believe I got this to work using a simple redirect on the webserver:

**Example for Firebase Hosting**
(other hosts would require different redirect setup)

1. Create your dynamic link like normal (without #).
`https://link.domain.com/share/sHoRtCoDe`
resolves to 
`https://www.domain.com/share?token=12345`

2. Setup firebase.json, notice redirect from `/share:params*` to `/#/share:params`
```json
{
  "hosting": [
    {
      "public": "build/web",
      "ignore": ["firebase.json", "**/.*", "**/node_modules/**"],
      "redirects": [
        {
          "source": "/share:params*",
          "destination": "https://www.domain.com/#/share:params",
          "type": 301
        }
      ],
      "rewrites": [
        { "source": "/share/**", "dynamicLinks": true },
        {
          "source": "**",
          "destination": "/index.html"
        }
      ],
      "appAssociation": "AUTO",
      "cleanUrls": true,
      "trailingSlash": false
    }
  ]
}
```

This setup will process a `/share` prefix dynamic link, redirect it to its long link or open in-app (depending on your link config), if it opens on the web server it will redirect from the non-hash version to the hash-version and will pass along URL params, Flutter web will then be able to read the full url path via `onGenerateRoute`. You can then parse it with `Uri.tryParse()`

**Comment by woutervanwijk** [comment]
A workaround might be a simple redirect using an html page. For logging in with email links, I setup the dynamic link to point to a new file, login.html. It logs into firebase without flutter and redirects to the app if login is ok. You might be able to do the same for all dynamic links. This is the login file:

```
<!DOCTYPE html>
<html>


<body>

  <!-- The core Firebase JS SDK is always required and must be listed first -->
  <script src="https://www.gstatic.com/firebasejs/7.5.0/firebase-app.js"></script>
  <script src="https://www.gstatic.com/firebasejs/7.5.0/firebase-auth.js"></script>
 
  <script>
    // Your web app's Firebase configuration
    var firebaseConfig = {
      apiKey: "",
      authDomain: "",
      databaseURL: "",
      projectId: "",
      storageBucket: "",
      messagingSenderId: "",
      appId: ""
    };
    // Initialize Firebase
    firebase.initializeApp(firebaseConfig);
  </script>

  <script>
  
// Confirm the link is a sign-in with email link.
if (firebase.auth().isSignInWithEmailLink(window.location.href)) {
  // Additional state parameters can also be passed via URL.
  // This can be used to continue the user's intended action before triggering
  // the sign-in operation.
  // Get the email if available. This should be available if the user completes
  // the flow on the same device where they started it.
  var email = window.localStorage.getItem('flutter.yourStoredEmailKey');
  // remove " "
  email = email.slice(1,-1);
  console.log('Email:', email, 'link', window.location.href);
  
  // Common errors could be invalid email and invalid or expired OTPs.
  if (!email) {
    // User opened the link on a different device. To prevent session fixation
    // attacks, ask the user to provide the associated email again. For example:
    email = window.prompt('Please provide your email for confirmation');
  }

  console.log('Email:', email, 'link', window.location.href);
  // The client SDK will parse the code from the link for you.
  firebase.auth().signInWithEmailLink(email, window.location.href)
    .then(function(result) {
      window.location = '/#/';
    })
    .catch(function(error) {
      // Some error occurred, you can inspect the code: error.code
      // Common errors could be invalid email and invalid or expired OTPs.
      alert(error);
    });
}

  </script>

</body>

</html>
```

**Comment by BrendonW** [comment]
> But I think the main issue with it is that is broken. If you press back button on browser while on initial/replaced page, the url deletes /#/ and after that you can navigate normally but url doesn't update anymore and you are unable to use back button again in any route. The current history stack is well saved internally by flutter because appbar is managing it well, but when pressing back on browser it doesn't work.

Flutter should NOT handle the history as a duplicate of the browser, it creates endless problems and I'm pretty sure it will never work well. I've seen folks go down that road before and madness is the end result. I cannot "train" my customers or their users to do anything, so users will use the back button on the browser and if it doesn't work "as-expected" then they will get confused, I'll get support calls, and happiness will decline.

Using the app-bar construct in the browser looks and behaves strangely, because users are treating the app as a website, not a mobile app. I do use the app-bar, but not to replace the browser navigation. I expect users to bookmark random app "urls" and also to share them, so they need to behave according to the user expectations. In the browser, I'd strongly prefer that flutter use the browser for all web browsing to the maximum extent possible. If I have to merge all server requests to pretend I'm seeing the "pre-hash" part of the URL, then that is trivial (I/m using golang). I just put my static route paths ahead of a catchall that can fall through to load Flutter.

**Comment by emmett-deen** [comment]
The above mentioned dynamic link issue is also applicable to redirects from api's using protocols like oauth 2.

**Comment by casetofon** [comment]
There's no way to implement a callback function in a OAuth2 flow if this doesn't work wihtout the # sign...

**Comment by jumper423** [comment]
Here comes the solution
https://github.com/flutter/flutter/pull/59797
https://github.com/flutter/engine/pull/19134

**Comment by iTomkinas** [comment]
There is a little problem with solution - last commit was 3 months ago. Sad that this issue is not getting too much attention

**Comment by franarolas** [comment]
> 
> 
> Here comes the solution
> #59797
> [flutter/engine#19134](https://github.com/flutter/engine/pull/19134)

Let's pray that this soultion gets accepted.

**Comment by mdebbar** [comment]
Hello everyone. Thanks for the feedback and comments. This feature will be available soon in the master channel.

Here are the steps to use it once it's available:

1. Add `<base href="/">` inside the `<head>` section of your `web/index.html` file.
2. Add the `flutter_web_plugins` dependency in `pubspec.yaml` if it doesn't already exist:
```yaml
dependencies:
  flutter_web_plugins:
    sdk: flutter
```
3. Add a `lib/configure_nonweb.dart` with the following content:
```dart
void configureApp() {
  // No-op.
}
```
4. Add a `lib/configure_web.dart` with the following content:
```dart
import 'package:flutter_web_plugins/flutter_web_plugins.dart';

void configureApp() {
  setUrlStrategy(PathUrlStrategy());
}
```
5. In `lib/main.dart`, do the following:
```dart
import 'package:flutter/material.dart';
import 'configure_nonweb.dart' if (dart.library.html) 'configure_web.dart';

void main() {
  configureApp();
  runApp(MyApp());
}
```

**Comment by rlee1990** [comment]
@md-weber any clue on when this will be available?

**Comment by mdebbar** [comment]
@rlee1990 I'm assuming you meant to tag me :)

The last PR remaining to close this issue is: https://github.com/flutter/flutter/pull/59797. Once that PR is merged, you should be able to follow the instructions above and get it working.

**Comment by rlee1990** [comment]
@mdebbar yes 🤦🏿‍♂️ and good thanks for the update.

**Comment by franarolas** [comment]
@mdebbar When should it be released to beta channel?

**Comment by rlee1990** [comment]
@mdebbar I am assuming that this has been added since I can now do what you mentioned above but now I get this error:
`TypeError: strategy.onPopState is not a function
    at _engine.MultiEntriesBrowserHistory.new._setupStrategy (http://localhost:56664/dart_sdk.js:144438:39)
    at _setupStrategy.next (<anonymous>)
    at runBody (http://localhost:56664/dart_sdk.js:37692:34)
    at Object._async [as async] (http://localhost:56664/dart_sdk.js:37723:7)
    at _engine.MultiEntriesBrowserHistory.new.[_setupStrategy] (http://localhost:56664/dart_sdk.js:144434:20)
    at _engine.MultiEntriesBrowserHistory.new.setLocationStrategy (http://localhost:56664/dart_sdk.js:144429:37)
    at setLocationStrategy.next (<anonymous>)
    at http://localhost:56664/dart_sdk.js:37672:33
    at _RootZone.runUnary (http://localhost:56664/dart_sdk.js:37526:58)
    at _FutureListener.thenAwait.handleValue (http://localhost:56664/dart_sdk.js:32500:29)
    at handleValueCallback (http://localhost:56664/dart_sdk.js:33047:49)
    at Function._propagateToListeners (http://localhost:56664/dart_sdk.js:33085:17)
    at _Future.new.[_completeWithValue] (http://localhost:56664/dart_sdk.js:32928:23)
    at async._AsyncCallbackEntry.new.callback (http://localhost:56664/dart_sdk.js:32950:35)
    at Object._microtaskLoop (http://localhost:56664/dart_sdk.js:37787:13)
    at _startMicrotaskLoop (http://localhost:56664/dart_sdk.js:37793:13)
    at http://localhost:56664/dart_sdk.js:33302:9
Error: Assertion failed:  (org-dartlang-sdk:///flutter_web_sdk/lib/_engine/engine/history.dart:193:12))
_hasSerialCount(currentState)
`

**Comment by franarolas** [comment]
@mdebbar I've been testing it for a few hours in realease mode, it works properly but if you try to access straight to the url, it just fails.
If I try to access a page straight from the browser, the router can't find it(Tested in a real server).
For example if you access "testweb.com" it works but if you access "testweb.com/loginpage" server returns 404 error, it doesn't happen with hashtag.

**Comment by creativecreatorormaybenot** [comment]
@franarolas You need to rewrite all paths to your `index.html` of course.

**Comment by rlee1990** [comment]
@creativecreatorormaybenot can you explain a bit more for if anyone needs this?

**Comment by franarolas** [comment]
> 
> 
> @franarolas You need to rewrite all paths to your `index.html` of course.

I didn't thought about that, thank you!

**Comment by mdebbar** [comment]
@franarolas the beta release process is explained here: https://github.com/flutter/flutter/wiki/Release-process

@creativecreatorormaybenot thanks for helping out :) @franarolas I had to make a few changes to the flutter web server (`flutter run ...`) to support this. When you deploy, you'd have to do something similar to your production server. Basically serve `index.html` for all unresolved paths.

**Comment by Pyciko** [comment]
Hey @mdebbar 
So how is it going? As I see, you commit was merged to the master branch. Should we follow your instructions or anything changed?

**Comment by emmett-deen** [comment]
Just wanted to point out that this seems to be on the beta channel now! Great work everyone 🍻

**Comment by Sun3** [comment]
As of Flutter version 1.24.0-4.0.pre.51 • channel master the trick is no longer working.

You now receive the following error:
```TypeError: dart.global._flutter_web_set_location_strategy is not a function```

**Comment by b34st80y** [comment]
@Sun3 You must switch to beta channel, not master channel

**Comment by KabluBR** [comment]
@mdebbar 
Just did what you said (5 steps) and I am receiving errors now:

TypeError: dart.global._flutter_web_set_location_strategy is not a function
    at Object.setUrlStrategy (http://localhost:27534/packages/flutter_web_plugins/src/navigation/js_url_strategy.dart.lib.js:243:17)
    at Object.configureApp (http://localhost:27534/packages/project/configure_web.dart.lib.js:10:18)


Can you help me on this? what I`d miss?

**Comment by mdebbar** [comment]
@Sun3 and @KabluBR could you please file new issues including `flutter doctor -v` output and the code of the app? Thanks!

**Comment by KabluBR** [comment]
@mdebbar 
Is very strange... I changed to beta channel and still not worked... 
sorry about flutter doctor -v, now I am sending...

TypeError: dart.global._flutter_web_set_location_strategy is not a function
    at Object.setUrlStrategy (http://localhost:29624/packages/flutter_web_plugins/src/navigation/js_url_strategy.dart.lib.js:243:17)
    at Object.configureApp (http://localhost:29624/packages/link_pagamento/configure_web.dart.lib.js:10:18)



flutter doctor -v
[√] Flutter (Channel master, 1.24.0-4.0.pre.123, on Microsoft Windows [Version 10.0.18363.1139], locale pt-BR)
    • Flutter version 1.24.0-4.0.pre.123 at c:\DEV\flutter
    • Framework revision 735f6eec2c (8 minutes ago), 2020-10-28 15:09:46 -0700
    • Engine revision 64e6599910
    • Dart version 2.11.0 (build 2.11.0-260.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\android-sdk-win
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = C:\android-sdk-win
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code, 64-bit edition (version 1.49.2)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 3.15.0

[√] Connected device (3 available)
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome 86.0.4240.111
    • Edge (web)       • edge       • web-javascript • Microsoft Edge 86.0.622.56

• No issues found!

**Comment by Sun3** [comment]
@mdebbar The following is my ```flutter doctor -v``` report:
```
[√] Flutter (Channel master, 1.24.0-4.0.pre.122, on Microsoft Windows [Version 10.0.19041.572], locale en-US)
    • Flutter version 1.24.0-4.0.pre.122 at C:\Users\marco\flutter
    • Framework revision 4aad058a49 (81 minutes ago), 2020-10-28 17:02:37 -0400
    • Engine revision 64e6599910
    • Dart version 2.11.0 (build 2.11.0-260.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at C:\Users\marco\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.7.6)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.7.30611.23
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code (version 1.50.1)
    • VS Code at C:\Users\marco\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.15.1

[√] Connected device (4 available)
    • Windows (desktop) • windows    • windows-x64    • Microsoft Windows [Version 10.0.19041.572]
    • Web Server (web)  • web-server • web-javascript • Flutter Tools
    • Chrome (web)      • chrome     • web-javascript • Google Chrome 86.0.4240.111
    • Edge (web)        • edge       • web-javascript • Microsoft Edge 86.0.622.51

• No issues found!
```

**Comment by creativecreatorormaybenot** [comment]
@KabluBR It would be great to have a **new** issue for this in order to keep issue tracking maintainable.

**Comment by rezajalilian** [comment]
> @mdebbar I've been testing it for a few hours in realease mode, it works properly but if you try to access straight to the url, it just fails.
> If I try to access a page straight from the browser, the router can't find it(Tested in a real server).
> For example if you access "testweb.com" it works but if you access "testweb.com/loginpage" server returns 404 error, it doesn't happen with hashtag.

Can you explain with an example? Thankful

**Comment by rezajalilian** [comment]
> > @franarolas You need to rewrite all paths to your `index.html` of course.
> 
> Silly of me I didn't thought about that, thank you!

Can you explain with an example? Thanks

**Comment by franarolas** [comment]
@KabluBR @Sun3 As @b34st80y said, you must switch to beta channel, your flutter doctor indicates that you are still in channel master.

@rezajalilian He is reffering to point the routes to the flutter web index.html, it is not related to Flutter.

**Comment by rezajalilian** [comment]
> @franarolas

 Excuse me, can you tell me what I should do now to avoid getting a 404 error? Do I have to do something on the server side? You can give a sample of your code

**Comment by Sun3** [comment]
Still not working for me...

**Comment by rezajalilian** [comment]
> > @KabluBR @Sun3 As @b34st80y said, you must switch to beta channel, your flutter doctor indicates that you are still in channel master.
> > @rezajalilian He is reffering to point the routes to the flutter web index.html, it is not related to Flutter.
> 
> @franarolas I had switched, as I have the same results I switched back to master, anyway it seems the fix is already on master by now...

please show example code or Article . Thanks

**Comment by adyking** [comment]
Hi I would like to know the changes on the index.html too please?

**Comment by franarolas** [comment]
> 
> 
> > @franarolas
> 
> Excuse me, can you tell me what I should do now to avoid getting a 404 error? Do I have to do something on the server side? You can give a sample of your code

It is server side, you can't control that in Flutter, you should try to point all routes to the index.html on your http server, usually apache or nginx.
I may implement that on the next days so as soon I've it I will share it here.

**Comment by creativecreatorormaybenot** [comment]
Usually Express.

**Comment by creativecreatorormaybenot** [comment]
Usually Express.js.

**Comment by Manjunath375** [comment]
> ## A quick fix for it:
> 1. Go to `main.dart.js` and look for following line:
>    `H.a(window.location.pathname)+H.a(window.location.search):"#"+a},`
> 2. Remove **#** in above line
> 3. Result:
>    `H.a(window.location.pathname)+H.a(window.location.search):""+a},`
> 
> This should give you https://example.com/ instead of https://example.com/#/

Can u tell where the exact line is located 

i got only below code in main.dart.js file

if(s==null)s=""
if(s.length===0||s==="#")return"/"
return C.d.cn(s,1)},
wd:function(a){return new P.o8([],[]).pG(window.history.state,!0)},
zh:function(a,b){var s,r
if(b.length===0){s=window.location.pathname
s.toString
r=window.location.search
r.toString
r=s+r
s=r}else s="#"+b
return s},

If i replace # with '' again i am getting the # in the url

**Comment by Manjunath375** [comment]
> Hello everyone. Thanks for the feedback and comments. This feature will be available soon in the master channel.
> 
> Here are the steps to use it once it's available:
> 
> 1. Add `<base href="/">` inside the `<head>` section of your `web/index.html` file. This will be added automatically for new projects created by `flutter create`. But for existing apps, the developer needs to add it manually.
> 2. Add the `flutter_web_plugins` dependency in `pubspec.yaml` if it doesn't already exist:
> 
> ```yaml
> dependencies:
>   flutter_web_plugins:
>     sdk: flutter
> ```
> 
> 1. Add a `lib/configure_nonweb.dart` with the following content:
> 
> ```dart
> void configureApp() {
>   // No-op.
> }
> ```
> 
> 1. Add a `lib/configure_web.dart` with the following content:
> 
> ```dart
> import 'package:flutter_web_plugins/flutter_web_plugins.dart';
> 
> void configureApp() {
>   setUrlStrategy(PathUrlStrategy());
> }
> ```
> 
> 1. In `lib/main.dart`, do the following:
> 
> ```dart
> import 'package:flutter/material.dart';
> import 'configure_nonweb.dart' if (dart.library.html) 'configure_web.dart';
> 
> void main() {
>   configureApp();
>   runApp(MyApp());
> }
> ```

in local it is running fine but when i push the code to server i am getting 404 error

**Comment by foxanna** [comment]
https://github.com/flutter/engine/pull/22371

**Comment by foxanna** [comment]
https://github.com/flutter/engine/pull/22371 was merged to `master` with https://github.com/flutter/flutter/pull/70163 so I assumed it was safe to switch to flutter master channel, apply recommended changes from https://github.com/flutter/flutter/issues/33245#issuecomment-705095853, and expect the web app to work without hash symbol.   

When run from Android Studio, it is indeed the case. The browser opens without `#` in the address bar (for example, `http://localhost:62007/`), and the navigation from the URL (for example, `http://localhost:62007/event/eventId`) works as expected.

However, when built with `flutter build web` and spun up in the local HTTP server, only the root navigation (for example, `http://127.0.0.1:8080/`) is working as expected, but navigating to a custom path (for example, `http://127.0.0.1:8080/event/eventId`) shows `HTTP ERROR 404`.

@mdebbar could you please update us on the current state of ability to customize the URL strategy for Flutter web apps? Is it too early to try a fix or steps to set it up have changed?

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 1.24.0-8.0.pre.194, on Mac OS X 10.15.7 19H2 darwin-x64, locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] VS Code (version 1.50.1)

• No issues found!

```

**Comment by creativecreatorormaybenot** [comment]
@foxanna You will have to use Express et al. in order to rewrite traffic from all paths to `index.html`. You cannot expect `/event/eventId` to open your Flutter app when you have configured your server in a way, where it only serves `index.html` for the root.

**Comment by KabluBR** [comment]
@foxanna I had same issue, to fix it in my case I managed to use only queryStrings...

So you can use a link like: http://127.0.0.1:8080/?eid=1234

Other way you can use htaccess and rewrite the url

But it would be nice to have a way to route all those arguments / routes.

**Comment by mdebbar** [comment]
@foxanna @KabluBR , like @creativecreatorormaybenot said, this is an issue with the local HTTP server that you are using. You have to configure it so that it loads the `web/index.html` file when you hit `/event/eventId` or any other path.

The built-in flutter web server does this automatically for you. But if you use a different server, you will need to figure out how to configure your server to do the same. The best place to ask questions like this is on StackOverflow. You'll find people who are familiar with the server you are using, and they might be able to help.

**Comment by devanflores** [comment]
Does anyone have any potential solutions to the 404 error? I have no idea where to start or what potential apis to look at

**Comment by devanflores** [comment]
@creativecreatorormaybenot what did you mean when you said you need to rewrite all paths to your index.html?

**Comment by taz1208** [comment]
> @foxanna @KabluBR , like @creativecreatorormaybenot said, this is an issue with the local HTTP server that you are using. You have to configure it so that it loads the `web/index.html` file when you hit `/event/eventId` or any other path.
> 
> The built-in flutter web server does this automatically for you. But if you use a different server, you will need to figure out how to configure your server to do the same. The best place to ask questions like this is on StackOverflow. You'll find people who are familiar with the server you are using, and they might be able to help.


Can you give an example? 
Please use golang if possible

**Comment by taz1208** [comment]
I found some solution.
```
package main
import (
  "net/http"
  "strings"
  "path"
  "os"
  // "log"
)

const FSPATH = "build/web/"

func main() {
    fs := http.FileServer(http.Dir(FSPATH))


    http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {

        if r.URL.Path != "/" {
            fullPath := FSPATH + strings.TrimPrefix(path.Clean(r.URL.Path), "/")
            _, err := os.Stat(fullPath)
            if err != nil {
                if !os.IsNotExist(err) {

                    panic(err)
                }
                // Requested file does not exist so we return the default (resolves to index.html)
                r.URL.Path = "/build/web/index.html"
            }
        }
        fs.ServeHTTP(w, r)
    })
    http.ListenAndServe(":3000", nil)
}
```
This is have some problem,
If remove hash(#) in URL path, all redirect to index.html 
It can't expect provide parameter like this "/book/1".

I can't use golang very well but I want someone helpful

**Comment by rahulmuthyam** [comment]
@taz1208  Regarding the how-to for redirection to index.html, please refer to this vue.js documentation: [Rewrite index.html](https://router.vuejs.org/guide/essentials/history-mode.html#example-server-configurations)
It has configurations for Apache, Nginx, Firebase hosting etc.

**Comment by sanalkv** [comment]
> Hello everyone. Thanks for the feedback and comments. This feature will be available soon in the master channel.
> 
> Here are the steps to use it once it's available:
> 
> 1. Add `<base href="/">` inside the `<head>` section of your `web/index.html` file. This will be added automatically for new projects created by `flutter create`. But for existing apps, the developer needs to add it manually.
> 2. Add the `flutter_web_plugins` dependency in `pubspec.yaml` if it doesn't already exist:
> 
> ```yaml
> dependencies:
>   flutter_web_plugins:
>     sdk: flutter
> ```
> 
> 1. Add a `lib/configure_nonweb.dart` with the following content:
> 
> ```dart
> void configureApp() {
>   // No-op.
> }
> ```
> 
> 1. Add a `lib/configure_web.dart` with the following content:
> 
> ```dart
> import 'package:flutter_web_plugins/flutter_web_plugins.dart';
> 
> void configureApp() {
>   setUrlStrategy(PathUrlStrategy());
> }
> ```
> 
> 1. In `lib/main.dart`, do the following:
> 
> ```dart
> import 'package:flutter/material.dart';
> import 'configure_nonweb.dart' if (dart.library.html) 'configure_web.dart';
> 
> void main() {
>   configureApp();
>   runApp(MyApp());
> }
> ```

this is working fine in debug but not in release. Please help. How to make it work it release build

**Comment by DannyJoinedApps** [comment]
> > Hello everyone. Thanks for the feedback and comments. This feature will be available soon in the master channel.
> > Here are the steps to use it once it's available:
> > 
> > 1. Add `<base href="/">` inside the `<head>` section of your `web/index.html` file. This will be added automatically for new projects created by `flutter create`. But for existing apps, the developer needs to add it manually.
> > 2. Add the `flutter_web_plugins` dependency in `pubspec.yaml` if it doesn't already exist:
> > 
> > ```yaml
> > dependencies:
> >   flutter_web_plugins:
> >     sdk: flutter
> > ```
> > 
> > 
> > 
> > 1. Add a `lib/configure_nonweb.dart` with the following content:
> > 
> > ```dart
> > void configureApp() {
> >   // No-op.
> > }
> > ```
> > 
> > 
> > 
> > 1. Add a `lib/configure_web.dart` with the following content:
> > 
> > ```dart
> > import 'package:flutter_web_plugins/flutter_web_plugins.dart';
> > 
> > void configureApp() {
> >   setUrlStrategy(PathUrlStrategy());
> > }
> > ```
> > 
> > 
> > 
> > 1. In `lib/main.dart`, do the following:
> > 
> > ```dart
> > import 'package:flutter/material.dart';
> > import 'configure_nonweb.dart' if (dart.library.html) 'configure_web.dart';
> > 
> > void main() {
> >   configureApp();
> >   runApp(MyApp());
> > }
> > ```
> 
> this is working fine in debug but not in release. Please help. How to make it work in release build

It's not working for me, even in debug. I do get the hash vanished, but as soon as I attempt to append the url with anything to get the router to fire I get a bunch of errors and no page:
TypeError: dart.global._flutter_web_set_location_strategy is not a function
    at Object.setUrlStrategy (http://localhost:56442/packages/flutter_web_plugins/src/navigation/js_url_strategy.dart.lib.js:281:17)
    at Object.configureApp (http://localhost:56442/packages/appname/configure_web.dart.lib.js:10:18)
    at main$ (http://localhost:56442/packages/appname/main.dart.lib.js:911:21)
    at main$.next (<anonymous>)
    at runBody (http://localhost:56442/dart_sdk.js:37976:34)
    at Object._async [as async] (http://localhost:56442/dart_sdk.js:38007:7)
    at main$ (http://localhost:56442/packages/appname/main.dart.lib.js:910:18)
    at main (http://localhost:56442/web_entrypoint.dart.lib.js:34:27)
    at main.next (<anonymous>)
    at http://localhost:56442/dart_sdk.js:37956:33

I need this because... I have a page which allows a user to reset their password, so I need the router to fire on receipt of a URL with settings, e.g. domain.com/passwordreset?oobcode=slkjdfljds?mode=password

I DID have this working before, even with the #. I could send users to a url including the has, e.g.:
domain.com/#/passwordreset?oobcode=slkjdfljds?mode=password

And it would work. But in the last few weeks google authentication did an update and my custom URL link configured in FireBase is getting wrecked. They changed something and now the # symbol is being interpreted as a placeholder and they insert the codes where the # is. This means I get urls in the email notifications which have the code in the wrong place and that in turn means that my flutter app's router doesn't get fired because the hash has been vanished.

I have contacted google about the breaking change to their authentication/emailtemplate/customlink issue. But if I could somehow just make the hash requirement vanish that would be great. Its jsut that if other companies (you know, like google, oh no flutter is google...) use hash's as special symbols when generating URLs, then flutter web apps expecting urls to have hashes in is going to be unpleasant. And I'm hosting on firebase by the way, so no URL rewriting on the server options for me.

**Comment by sanalkv** [comment]
> > > Hello everyone. Thanks for the feedback and comments. This feature will be available soon in the master channel.
> > > Here are the steps to use it once it's available:
> > > 
> > > 1. Add `<base href="/">` inside the `<head>` section of your `web/index.html` file. This will be added automatically for new projects created by `flutter create`. But for existing apps, the developer needs to add it manually.
> > > 2. Add the `flutter_web_plugins` dependency in `pubspec.yaml` if it doesn't already exist:
> > > 
> > > ```yaml
> > > dependencies:
> > >   flutter_web_plugins:
> > >     sdk: flutter
> > > ```
> > > 
> > > 
> > > 
> > > 1. Add a `lib/configure_nonweb.dart` with the following content:
> > > 
> > > ```dart
> > > void configureApp() {
> > >   // No-op.
> > > }
> > > ```
> > > 
> > > 
> > > 
> > > 1. Add a `lib/configure_web.dart` with the following content:
> > > 
> > > ```dart
> > > import 'package:flutter_web_plugins/flutter_web_plugins.dart';
> > > 
> > > void configureApp() {
> > >   setUrlStrategy(PathUrlStrategy());
> > > }
> > > ```
> > > 
> > > 
> > > 
> > > 1. In `lib/main.dart`, do the following:
> > > 
> > > ```dart
> > > import 'package:flutter/material.dart';
> > > import 'configure_nonweb.dart' if (dart.library.html) 'configure_web.dart';
> > > 
> > > void main() {
> > >   configureApp();
> > >   runApp(MyApp());
> > > }
> > > ```
> > 
> > 
> > this is working fine in debug but not in release. Please help. How to make it work in release build
> 
> It's not working for me, even in debug. I do get the hash vanished, but as soon as I attempt to append the url with anything to get the router to fire I get a bunch of errors and no page:
> TypeError: dart.global._flutter_web_set_location_strategy is not a function
> at Object.setUrlStrategy (http://localhost:56442/packages/flutter_web_plugins/src/navigation/js_url_strategy.dart.lib.js:281:17)
> at Object.configureApp (http://localhost:56442/packages/appname/configure_web.dart.lib.js:10:18)
> at main$ (http://localhost:56442/packages/appname/main.dart.lib.js:911:21)
> at main$.next ()
> at runBody (http://localhost:56442/dart_sdk.js:37976:34)
> at Object._async [as async] (http://localhost:56442/dart_sdk.js:38007:7)
> at main$ (http://localhost:56442/packages/appname/main.dart.lib.js:910:18)
> at main (http://localhost:56442/web_entrypoint.dart.lib.js:34:27)
> at main.next ()
> at http://localhost:56442/dart_sdk.js:37956:33
> 
> I need this because... I have a page which allows a user to reset their password, so I need the router to fire on receipt of a URL with settings, e.g. domain.com/passwordreset?oobcode=slkjdfljds?mode=password
> 
> I DID have this working before, even with the #. I could send users to a url including the has, e.g.:
> domain.com/#/passwordreset?oobcode=slkjdfljds?mode=password
> 
> And it would work. But in the last few weeks google authentication did an update and my custom URL link configured in FireBase is getting wrecked. They changed something and now the # symbol is being interpreted as a placeholder and they insert the codes where the # is. This means I get urls in the email notifications which have the code in the wrong place and that in turn means that my flutter app's router doesn't get fired because the hash has been vanished.
> 
> I have contacted google about the breaking change to their authentication/emailtemplate/customlink issue. But if I could somehow just make the hash requirement vanish that would be great. Its jsut that if other companies (you know, like google, oh no flutter is google...) use hash's as special symbols when generating URLs, then flutter web apps expecting urls to have hashes in is going to be unpleasant. And I'm hosting on firebase by the way, so no URL rewriting on the server options for me.

TypeError: dart.global._flutter_web_set_location_strategy is not a function - this issue will be resolved when you run your project on Channel beta, 1.23.0-18.1.pre .

**Comment by bojidartonchev** [comment]
I also had the same issue using `setUrlStrategy(PathUrlStrategy());` **in release**.

These steps solved it for me:

1. Create `.htaccess` file at the level of your `index.html`
2. Add the following content:
```
<IfModule mod_rewrite.c>
  RewriteEngine On
  RewriteBase /
  RewriteRule ^index\.html$ - [L]
  RewriteCond %{REQUEST_FILENAME} !-f
  RewriteCond %{REQUEST_FILENAME} !-d
  RewriteRule . /index.html [L]
</IfModule>
```

**Comment by creativecreatorormaybenot** [comment]
I made a package that simplifies the process 🚀

You can use `url_strategy` like this:

```dart
import 'package:url_strategy/url_strategy.dart';

void main() {
  // Here we set the URL strategy for our web app.
  // It is safe to call this function when running on mobile or desktop as well.
  setPathUrlStrategy();
  runApp(MyApp());
}
```

https://github.com/simpleclub/url_strategy

**Comment by jamesblasco** [comment]
Thank you @creativecreatorormaybenot! I personally think this should be the default way to go and be implemented like this inside the framework

**Comment by rlee1990** [comment]
@creativecreatorormaybenot will this work with auto_router package without issues? I had issues using the configure app approach where it would not work will with keeping the with state intact for using the auth guards.

**Comment by rlee1990** [comment]
@creativecreatorormaybenot from my test it did not work.

**Comment by creativecreatorormaybenot** [comment]
@rlee1990 It does work. It is simply a simpler version of the manual method. I am not sure what kind of setup you have going and what kind of problem you are experiencing. You might find help by asking about your specific problem on StackOverflow 🙂

**Comment by rlee1990** [comment]
@creativecreatorormaybenot What I mean is it does not work with the Auto_router package. Not stating it's on your end been working with the package creator to see what's wrong if it's on their in or just flutter in general

**Comment by keitoro** [comment]
It is a nice workaround but it seems you can't link directly to pages as it needs to go through index.html. That limits its use significantly.

**Comment by mdebbar** [comment]
@keitoro if you are running the app using `flutter run ...` it should work fine. You can link directly to `http://localhost:12345/page/1` for example.

If you are deploying the app to production, you need to setup your web server so that it serves `index.html` on all unknown paths.

**Comment by speller** [comment]
In the most recent flutter stable, functions `setUrlStrategy` and `PathUrlStrategy` are undefined even after adding the `flutter_web_plugins` dependency.

**Comment by creativecreatorormaybenot** [comment]
@speller Of course (`stable` has no web support).

See [my answer](https://stackoverflow.com/a/65709246/6509751) for a way to still be able to run your mobile app on `stable`.

**Comment by speller** [comment]
@creativecreatorormaybenot Ah, sorry, I forgot that I've enabled web support on the stable channel manually because I had stability issues on beta but I still need to develop the web version (which runs well so far on my app).

**Comment by nxgenpro** [comment]
> > Hello everyone. Thanks for the feedback and comments. This feature will be available soon in the master channel.
> > Here are the steps to use it once it's available:
> > 
> > 1. Add `<base href="/">` inside the `<head>` section of your `web/index.html` file. This will be added automatically for new projects created by `flutter create`. But for existing apps, the developer needs to add it manually.
> > 2. Add the `flutter_web_plugins` dependency in `pubspec.yaml` if it doesn't already exist:
> > 
> > ```yaml
> > dependencies:
> >   flutter_web_plugins:
> >     sdk: flutter
> > ```
> > 
> > 
> > 
> > 1. Add a `lib/configure_nonweb.dart` with the following content:
> > 
> > ```dart
> > void configureApp() {
> >   // No-op.
> > }
> > ```
> > 
> > 
> > 
> > 1. Add a `lib/configure_web.dart` with the following content:
> > 
> > ```dart
> > import 'package:flutter_web_plugins/flutter_web_plugins.dart';
> > 
> > void configureApp() {
> >   setUrlStrategy(PathUrlStrategy());
> > }
> > ```
> > 
> > 
> > 
> > 1. In `lib/main.dart`, do the following:
> > 
> > ```dart
> > import 'package:flutter/material.dart';
> > import 'configure_nonweb.dart' if (dart.library.html) 'configure_web.dart';
> > 
> > void main() {
> >   configureApp();
> >   runApp(MyApp());
> > }
> > ```
> 
> in local it is running fine but when i push the code to server i am getting 404 error. Need proper solution on how to work onremoving # from url

hey, did you find any solution for this?

**Comment by arualana** [comment]
@nxgenpro https://stackoverflow.com/a/36623117/3808307

**Comment by SameerChorge94** [comment]
> > Hello everyone. Thanks for the feedback and comments. This feature will be available soon in the master channel.
> > Here are the steps to use it once it's available:
> > 
> > 1. Add `<base href="/">` inside the `<head>` section of your `web/index.html` file. This will be added automatically for new projects created by `flutter create`. But for existing apps, the developer needs to add it manually.
> > 2. Add the `flutter_web_plugins` dependency in `pubspec.yaml` if it doesn't already exist:
> > 
> > ```yaml
> > dependencies:
> >   flutter_web_plugins:
> >     sdk: flutter
> > ```
> > 
> > 
> > 
> > 1. Add a `lib/configure_nonweb.dart` with the following content:
> > 
> > ```dart
> > void configureApp() {
> >   // No-op.
> > }
> > ```
> > 
> > 
> > 
> > 1. Add a `lib/configure_web.dart` with the following content:
> > 
> > ```dart
> > import 'package:flutter_web_plugins/flutter_web_plugins.dart';
> > 
> > void configureApp() {
> >   setUrlStrategy(PathUrlStrategy());
> > }
> > ```
> > 
> > 
> > 
> > 1. In `lib/main.dart`, do the following:
> > 
> > ```dart
> > import 'package:flutter/material.dart';
> > import 'configure_nonweb.dart' if (dart.library.html) 'configure_web.dart';
> > 
> > void main() {
> >   configureApp();
> >   runApp(MyApp());
> > }
> > ```
> 
> this is working fine in debug but not in release. Please help. How to make it work in release build

Hii Any update on this issue , facing same problem

**Comment by Ucoon** [comment]
opening subpage in browser throws 404 error，is there any other soluction?

**Comment by ronytesler** [comment]
Why is the issued closed? Is this working properly? https://flutter.dev/docs/development/ui/navigation/url-strategies#configuring-the-url-strategy

**Comment by jakusb** [comment]
> I also had the same issue using `setUrlStrategy(PathUrlStrategy());` **in release**.
> 
> These steps solved it for me:
> 
> 1. Create `.htaccess` file at the level of your `index.html`
> 2. Add the following content:
> 
> ```
> <IfModule mod_rewrite.c>
>   RewriteEngine On
>   RewriteBase /
>   RewriteRule ^index\.html$ - [L]
>   RewriteCond %{REQUEST_FILENAME} !-f
>   RewriteCond %{REQUEST_FILENAME} !-d
>   RewriteRule . /index.html [L]
> </IfModule>
> ```

For me it als required Directory tags around it. And I added it directly to Apache config. Works perfectly now! Thanks!

**Comment by gauravmehta13** [comment]
anything on this yet?
